cpmpy/transformations/linearize.py

```diff
@@ -5,190 +5,187 @@
-        else:
-            cons = lhs - 1 >= rhs
-        cons = flatten_constraint(cons)
-        cons = linearize_constraint(cons)
-        return cons
-
-    if cpm_expr.name == "!=":
-        # TODO: this can be optimized, see https://github.com/CPMpy/cpmpy/pull/93/files#r824163315
-        lhs, rhs = cpm_expr.args
-        # Special case: BV != BV
-        if isinstance(lhs, _BoolVarImpl) and isinstance(rhs, _BoolVarImpl):
-            return [lhs + rhs == 1]
-        # Normal case: big M implementation
-        z = boolvar()
-
-        # Calculate bounds of M = |lhs - rhs| + 1
-        bound1, _ = get_or_make_var(1 + lhs - rhs)
-        bound2, _ = get_or_make_var(1 + rhs - lhs)
-        M = max(bound1.ub, bound2.ub)
-
-        if lhs.name == "sum":
-            lhs = Operator("wsum", [[1]*len(lhs.args), lhs.args])
-            cons_lhs = []
-        if _wsum_should(lhs):
-            lhs, cons_lhs = Operator("wsum",_wsum_make(lhs)), []
-        if lhs.name != "wsum":
-            lhs, cons_lhs = get_or_make_var(lhs)
-        # Rewrite of constraints:
-        #   c1 = lhs - rhs <= Mz - 1
-        #   c2 = lhs - rhs >= -M(1-z) + 1
-        c1 = Operator("wsum", [[-M, 1],       [z, 1]]) + lhs <= rhs      # TODO: this can be optimized, see https://github.com/CPMpy/cpmpy/issues/97
-        c2 = Operator("wsum", [[-M, (M - 1)], [z, 1]]) + lhs >= rhs      # TODO: this can be optimized, see https://github.com/CPMpy/cpmpy/issues/97
-
-        return linearize_constraint(flatten_constraint(cons_lhs)) + [c1, c2]
-
-    if cpm_expr.name in [">=", "<=", "=="] and \
-            isinstance(cpm_expr.args[0], Operator) and cpm_expr.args[0].name == "mul":
-        lhs, rhs = cpm_expr.args
-        if len(lhs.args) == 2:
-            # multiplication of var and constant
-            if _wsum_should(lhs):
-                return [Comparison(cpm_expr.name, Operator("wsum", _wsum_make(lhs)), rhs)]
-            return [cpm_expr]
+            # BV -> LinExpr
+            if isinstance(cond, _BoolVarImpl):
+                lin_sub = linearize_constraint(sub_expr, supported=supported, reified=True)
+                return [cond.implies(lin) for lin in lin_sub]
 
+    # comparisons
+    if isinstance(cpm_expr, Comparison):
         lhs, rhs = cpm_expr.args
-        var, constraints = get_or_make_var(lhs.args[0] * lhs.args[1])
-        # decompose long multiplication in several multiplications with two args
-        for arg in lhs.args[2:]:
-            var, cons = get_or_make_var(var * arg)
-            constraints += cons
 
-        return constraints + [Comparison(cpm_expr.name, var, rhs)]
+        if isinstance(lhs, _NumVarImpl) and isinstance(rhs, _NumVarImpl):
+            # bring numvar to lhs
+            lhs = 1 * lhs + -1 * rhs
+            rhs = 0
+
+        # linearize unsupported operators
+        elif isinstance(lhs, Operator) and lhs.name not in supported: # TODO: add mul, (abs?), (mod?), (pow?)
```

Comment:
mul, div, mod, pow are not linearizable unless in special circumstances (e.g., div/mul by constant). `abs` will be fixed as a global constraint.

Reply:
Here's what I would do with `x mod y` and `x div y`: introduce the equality `x = y*[x div y] + [x mod y]` for auxiliary symbols `[x div y]` and `[x mod y]` (with appropriate bounds). Then this will simplify to the `*` and `+` operators, which in linearize will just check that `y` is a constant, which is what we want.

Reply:
yes

Reply:
So, TODO for later? Ok if I make an issue?

Reply:
This can indeed be added in a later iteration. Ok to leave it for now and create an issue